File: kaji/ts/src/cli/add.ts
/** `kaji add`: rollback-safe copied integration bundle installation. */
import { join, resolve } from "node:path";

import {
  BundleTransitionError,
  classifyIntegrationBundle,
  installIntegrationBundle,
  type BundleStatus,
} from "@/cli/integration-copy";
import {
  formatIntegrationError,
  IntegrationExperimentalError,
  loadManifest,
  loadRegistryIndex,
  type LoadedIntegrationManifest,
  type RegistryIndexDocument,
} from "@/integrations/registry-loader";

export interface AddOptions {
  readonly registryRoot: string;
  readonly schemaRoot?: string;
  readonly log?: (message: string) => void;
}

interface Args {
  readonly name: string;
  readonly out?: string;
  readonly force: boolean;
  readonly allowExperimental: boolean;
  readonly check: boolean;
  readonly json: boolean;
}

function parseArgs(argv: string[], log: (message: string) => void): Args | undefined {
  const name = argv[0];
  if (name === undefined || name.startsWith("-")) {
    log("usage: kaji add <name> [--out <dir>] [--force] [--allow-experimental] [--check] [--json]");
    return undefined;
  }
  let out: string | undefined;
  let force = false;
  let allowExperimental = false;
  let check = false;
  let json = false;
  for (let index = 1; index < argv.length; index += 1) {
    const argument = argv[index]!;
    if (argument === "--out") {
      const value = argv[++index];
      if (value === undefined || value.startsWith("--")) {
        log("--out requires a value");
        return undefined;
      }
      out = value;
    } else if (argument === "--force") {
      force = true;
    } else if (argument === "--allow-experimental") {
      allowExperimental = true;
    } else if (argument === "--check") {
      check = true;
    } else if (argument === "--json") {
      json = true;
    } else {
      log(`Unknown argument: ${argument}`);
      return undefined;
    }
  }
  if (check && force) {
    log("--check cannot be combined with --force");
    return undefined;
  }
  return { name, out, force, allowExperimental, check, json };
}

function exitCode(status: BundleStatus): number {
  return { current: 0, absent: 3, outdated: 4, modified: 5, demoted: 6 }[status.state];
}

function shellQuote(value: string): string {
  return /^[A-Za-z0-9_./:@+-]+$/u.test(value) ? value : `'${value.replaceAll("'", `'"'"'`)}'`;
}

function nextCommand(status: BundleStatus, manifest: LoadedIntegrationManifest): string {
  const command = ["bunx", "--package", "@kaji/sdk", "kaji", "add", manifest.name];
  if (manifest.stability === "experimental") command.push("--allow-experimental");
  command.push("--out", status.destination);
  if (status.state === "outdated") command.push("--force");
  else if (status.state !== "absent") command.push("--check");
  return command.map(shellQuote).join(" ");
}

function renderStatus(
  status: BundleStatus,
  manifest: LoadedIntegrationManifest,
  json: boolean,
  log: (message: string) => void,
): void {
  const next_command = nextCommand(status, manifest);
  if (json) {
    log(
      JSON.stringify({
        state: status.state,
        integration: manifest.name,
        runtime: "typescript",
        destination: status.destination,
        reason_code: status.reasonCode,
        next_command,
      }),
    );
    return;
  }
  log(`${status.state}: ${manifest.name} at ${status.destination} (${status.reasonCode})`);
  log(`next: ${next_command}`);
}

function setupGuidance(manifest: LoadedIntegrationManifest, log: (message: string) => void): void {
  if (manifest.name === "github" && manifest.auth.kind === "env") {
    log("next: set GITHUB_TOKEN to a fine-grained token limited to the configured repositories");
    log(`docs: ${manifest.auth.docs}`);
  } else if (manifest.auth.kind === "env") {
    log(`next: set ${manifest.auth.env} in your environment`);
  }
}

export async function add(argv: string[], opts: AddOptions): Promise<number> {
  const log = opts.log ?? ((message: string) => console.log(message));
  const args = parseArgs(argv, log);
  if (args === undefined) return argv.includes("--check") && argv.includes("--force") ? 2 : 1;

  let index: RegistryIndexDocument;
  try {
    index = await loadRegistryIndex(opts.registryRoot, { schemaRoot: opts.schemaRoot });
  } catch (error) {
    log(formatIntegrationError(error));
    return 1;
  }
  const entry = index.integrations[args.name];
  if (entry === undefined) {
    const available = Object.keys(index.integrations).sort().join(", ") || "(none)";
    log(`Unknown integration: '${args.name}'. Available: ${available}`);
    return 1;
  }
  if (entry.stability === "experimental" && !args.allowExperimental && !args.check) {
    log(formatIntegrationError(new IntegrationExperimentalError(args.name)));
    return 1;
  }

  let manifest: LoadedIntegrationManifest;
  try {
    manifest = await loadManifest(opts.registryRoot, args.name, {
      schemaRoot: opts.schemaRoot,
      index,
    });
  } catch (error) {
    log(formatIntegrationError(error));
    return 1;
  }
  const destination = resolve(args.out ?? join("./integrations", args.name));
  const context = { manifest, entry, destination, runtime: "typescript" as const };
  if (args.check) {
    try {
      const status = await classifyIntegrationBundle(context);
      renderStatus(status, manifest, args.json, log);
      return exitCode(status);
    } catch (error) {
      log(error instanceof Error ? error.message : "Integration check failed");
      return 1;
    }
  }

  let status: BundleStatus;
  try {
    status = await installIntegrationBundle({ ...context, force: args.force });
  } catch (error) {
    if (error instanceof BundleTransitionError) {
      renderStatus(error.status, manifest, args.json, log);
      return exitCode(error.status);
    }
    log(error instanceof Error ? error.message : "Integration copy failed");
    return 1;
  }
  if (args.json) {
    renderStatus(status, manifest, true, log);
    return 0;
  }
  if (status.written.length > 0)
    log(`Wrote ${status.written.length} file(s) to ${status.destination}`);
  else log(`Current integration: ${manifest.name} at ${status.destination}`);
  setupGuidance(manifest, log);
  return 0;
}

File: kaji/packages/py/src/cli/add.py
"""`kaji add <name>` -- install an integration from the registry.

shadcn-style: copies the integration's source files into the user's
project so they own and can edit the copies.
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path
import shlex
import sys

from kaji.integrations import (
    IntegrationNotFound,
    ManifestError,
    list_integrations,
    load_manifest,
)
from kaji.integrations.copy import (
    BundleStatus,
    BundleTransitionError,
    classify_integration_bundle,
    install_integration_bundle,
)

from ._pkg import TYPESCRIPT_SDK_CLI
from ._style import color


ADD_USAGE = (
    "usage: kaji add <name> [--out <dir>] [--force] "
    "[--allow-experimental] [--check] [--json]"
)


def add_parser(sub: argparse._SubParsersAction) -> None:
    p = sub.add_parser(
        "add",
        help="install an integration into your project (shadcn-style copy)",
    )
    p.add_argument("name", help="integration name (see `kaji list-integrations`)")
    p.add_argument(
        "--out",
        help="destination directory (default: ./integrations/<name>)",
    )
    p.add_argument(
        "--force", action="store_true", help="replace an unmodified outdated bundle"
    )
    p.add_argument(
        "--allow-experimental",
        action="store_true",
        help="allow copying an experimental integration",
    )
    p.add_argument("--check", action="store_true", help="classify without writing")
    p.add_argument("--json", action="store_true", help="print closed JSON output")
    p.set_defaults(func=run)


def _next_command(status: BundleStatus, manifest_name: str, experimental: bool) -> str:
    command = ["python", "-m", "kaji.cli", "add", manifest_name]
    if experimental:
        command.append("--allow-experimental")
    command.extend(("--out", str(status.destination)))
    if status.state == "outdated":
        command.append("--force")
    elif status.state not in {"absent"}:
        command.append("--check")
    return shlex.join(command)


def _render_status(
    status: BundleStatus, manifest_name: str, experimental: bool, json_output: bool
) -> None:
    next_command = _next_command(status, manifest_name, experimental)
    if json_output:
        print(
            json.dumps(
                {
                    "state": status.state,
                    "integration": manifest_name,
                    "runtime": "python",
                    "destination": str(status.destination),
                    "reason_code": status.reason_code,
                    "next_command": next_command,
                },
                separators=(",", ":"),
            )
        )
        return
    print(
        f"{status.state}: {manifest_name} at {status.destination} ({status.reason_code})"
    )
    print(f"next: {next_command}")


def run(args: argparse.Namespace) -> int:
    name = args.name
    dest = (Path(args.out) if args.out else Path("./integrations") / name).absolute()
    if args.check and args.force:
        print(
            color("--check cannot be combined with --force", "red"),
            file=sys.stderr,
        )
        print(ADD_USAGE, file=sys.stderr)
        return 2
    try:
        manifest = load_manifest(name)
    except IntegrationNotFound:
        available = ", ".join(list_integrations()) or "(none)"
        print(color(f"Unknown integration: {name!r}.", "red"), file=sys.stderr)
        print(f"Available: {available}", file=sys.stderr)
        return 1
    except ManifestError as e:
        print(color(f"Manifest error: {e}", "red"), file=sys.stderr)
        return 1

    if manifest.stability == "experimental" and not (
        args.allow_experimental or args.check
    ):
        print(
            color(
                f"Integration {name!r} is experimental. Re-run with --allow-experimental.",
                "red",
            ),
            file=sys.stderr,
        )
        return 1

    if args.check:
        status = classify_integration_bundle(manifest, dest, runtime="python")
        _render_status(
            status, manifest.name, manifest.stability == "experimental", args.json
        )
        return status.exit_code

    try:
        status = install_integration_bundle(
            manifest, dest, runtime="python", force=args.force
        )
    except BundleTransitionError as error:
        _render_status(
            error.status,
            manifest.name,
            manifest.stability == "experimental",
            args.json,
        )
        return error.status.exit_code
    except ManifestError as e:
        print(color(f"Install error: {e}", "red"), file=sys.stderr)
        return 1

    if args.json:
        _render_status(
            status, manifest.name, manifest.stability == "experimental", True
        )
        return 0
    if not status.written:
        _render_status(
            status, manifest.name, manifest.stability == "experimental", False
        )
        return 0
    for p in status.written:
        print(f"  wrote {p}")
    print()
    print(color(f"Installed integration: {manifest.name} v{manifest.version}", "green"))
    print(f"  {manifest.description}")
    if manifest.auth.kind == "env" and manifest.auth.env:
        if manifest.name == "github":
            print(
                color(
                    f"next: set {manifest.auth.env} to a fine-grained token limited to the configured repositories",
                    "yellow",
                )
            )
            print(f"docs: {manifest.auth.docs}")
        else:
            msg = f"  next: set {manifest.auth.env} in your environment"
            if manifest.auth.docs:
                msg += f" (see {manifest.auth.docs})"
            print(color(msg, "yellow"))
    elif manifest.auth.kind == "oauth":
        scopes = ", ".join(manifest.auth.scopes) or "(none declared)"
        print(color(f"  client ID env: {manifest.auth.client_id_env}", "yellow"))
        if manifest.auth.client_secret_env:
            print(
                color(
                    f"  client secret env: {manifest.auth.client_secret_env}", "yellow"
                )
            )
        print(color(f"  scopes: {scopes}", "yellow"))
        if manifest.auth.docs:
            print(f"  docs: {manifest.auth.docs}")
        print(
            "  connect (Python): "
            f"python -m kaji.cli connect {manifest.name} "
            "--principal <stable-host-principal-id>"
        )
        print(
            "  connect (TypeScript): "
            f"{TYPESCRIPT_SDK_CLI} connect {manifest.name} "
            "--principal <stable-host-principal-id>"
        )
    if manifest.extras and manifest.auth.kind != "oauth":
        print(color(f"  also: pip install {' '.join(manifest.extras)}", "yellow"))
    return 0
